### ISSUE TITLE
Check reshuffling matrices

### ISSUE THREAD

**Issue by rdemaria** [SixTrack/SixTrack#251]
Check whether matrices put in the header of binary files are always identical for each particle (both in 6D tracking and 4D with energy scan) or they are reshuffled correctly for particle loss.

**Comment by kyrsjo** [comment]
The matrix which is written to the header of fort.90 etc. is `tas(ia,i,j)` where `ia` is the particle index; the header is written once per pair (I expect each pair to have indentical matrices). The matrix has has dimension `(npart,6,6)`.

The `tas` matrix gets set quite early in `program maincr`, in two different code paths.

In the first path `if (iclo6.eq.1.or.iclo6.eq.2) then` it is copied from `tasm(6,6)` to (what I think is) all "particle" positions in `tas(ib3,6,6)`.
Here `iclo6` is read from the TRAC block and decides whether or not to calculate the 6D closed orbit using the DA package; setting 1 means calculate it, setting 2 means calculate and add to the initial coordinates. However there are several combinations of settings in e.g. DIFF that will override and disable iclo6.
The loop over particles is actually written as 
```
            do 190 ib1=1,napx
              ib3=ib1+(m+ib-2)*napx
```
where `ib3` is used as the particle index for `clo6v`, `clopcv`,`di0xs`, `di0zs`, `dip0xs`, `dip0zs`, `qwcs`, and `tas`.
Here `ib` is a loop variable running from 1 to `imc`, just outside the selection of code paths, and `m` from 1 to `mmac` quite a bit above that again.

In the second code path (`else`), it sets the same array, this time with the particle loop being
```
            do 170 i=1,napx
              iar=(m+ib-2)*napx+i
```

Note that just above the outermost `if` for the two code paths, we have `if(idp.ne.1.or.iation.ne.1) iclo6=0`.

A bit below the setting of `tas` etc. through the two code paths, there is the line `napx=(napx*imc)*mmac`. Here `imc` is read from the TRAC block in fort.3 and (according to the manual) defined as:
> Number of variations of the relative momentum deviation the relative momentum deviation ∆p/p_0.
> The maximum value of is taken from that of the first particle in the Initial Coordinates block (3.6.2).
> The variation will be between ± ∆p/p_0 (max) in steps of ∆p/p_0 (max) / (imc–1 ).
It is however set =1 in the DIFF block if ndum=1, or if the CORR block is active in fort.3.
The `imc` variable is not modified anywhere else.
The `mmac` flag is set in the FLUC block, however it is there fixed to 1 since it cannot be greater than the parameter `nmac` which is 1.

The code then initializes the particle arrays (`SUBROUTINE 'ANFB' IN-LINE`) using the `tas` etc. for each particle pair (`do 340 ia=1,napx,2`), then writes the headers for fort.90 etc.

To me, this indicates that the values of `tas` ARE different for each particle (pair).

**Comment by kyrsjo** [comment]
Survey of arrays with dimension `npart` (bnlelens/rhicelens code excluded, variables related to C/R excluded):

Allocatable tables from `datamods`:
 * `al(6,2,npart,nele)`
   * Read in `+cd thcklin` - used for kick in thck6d when `ktrack=1`
   * Updated in `subroutine envarsv`
 * `as(6,2,npart,nele)`
   * Read in `+cd thcklin` - used in thck6d
   * Updated in `subroutine envarsv`
 * `ekv(npart,nele)`
   * Copy of `ek(:)` (in `program maincr`)
 * `hv(6,2,npart,nblo)`
   * In `program maincr`, just after calling envarsv, `hv=al` (and then modified); then used to set `bl1v`.
 * `bl1v(6,2,npart,nblo)`
   * in subroutine thin4d, used for kick in thick4d when `ktrack=1`

In `+cd beamdim` (no common):
```
       crkveb(npart),cikveb(npart),rho2b(npart),tkb(npart),    &
     &r2b(npart),rb(npart),rkb(npart),                                  &
     &xrb(npart),zrb(npart),xbb(npart),zbb(npart),crxb(npart),          &
     &crzb(npart),cbxb(npart),cbzb(npart)
```

In `+cd common`: `track6d(6,npart)`
In `+cd commonmn`: `fokqv(npart)`, `nnumxv(npart)`, `xsv(npart)`, `zsc(npart)`, `xv(2,npart)`, `yv(2,npart)`, `dam(npart)`, `ekkv(npart)`, `sigmv(npart)`, `dpsv(npart)`, `dp0v(npart)`, `sigmv6(npart)`, dpsv6(npart)`, `ejv(npart)`, `ejfv(npart)`, xlv(npart)`, `zlv(npart)`, `pstop(npart)`, `rvv(npart)`, (if `rvet` : `rvet(npart)`), `ejf0v(npart)`, `numxv(npart)`, `nms(npart)`, `nlostp(npart)`;
```
 dpd(npart),dpsq(npart),fok(npart),rho(npart),       &
     &fok1(npart),si(npart),co(npart),g(npart),gl(npart),sm1(npart),    &
     &sm2(npart),sm3(npart),sm12(npart),as3(npart),as4(npart),          &
     &as6(npart),sm23(npart),rhoc(npart),siq(npart),aek(npart),         &
     &afok(npart),hp(npart),hm(npart),hc(npart),hs(npart),wf(npart),    &
     &wfa(npart),wfhi(npart),rhoi(npart),hi(npart),fi(npart),hi1(npart),&
     &xvl(2,npart),yvl(2,npart),ejvl(npart),dpsvl(npart),oidpsv(npart), &
     &sigmvl(npart),iv(npart),aperv(npart,2),ixv(npart),clov(2,npart),  &
     &clopv(2,npart),alf0v(npart,2),bet0v(npart,2),ampv(npart)

       clo6v(3,npart),clop6v(3,npart),                     &
     &tas(npart,6,6),qwcs(npart,3),di0xs(npart),  
       di0zs(npart),dip0xs(npart),dip0zs(npart)
        tasau(npart,6,6)
```
In `+cd commontr`: `dpsv1(npart)`

In dbtrthin:
```xineff(npart),yineff(npart),xpineff(npart),ypineff(npart)```

In dbthin6d
```ipart(npart), flukaname(npart) cx(npart),cxp(npart),cy(npart),cyp(npart),       &
     &cp(npart),cs(npart),rcx(npart),rcxp(npart),rcy(npart),rcyp(npart),&
     &rcp(npart),rcs(npart),rcx0(npart),rcxp0(npart),rcy0(npart),       &
     &rcyp0(npart),rcp0(npart) , xgrd(npart),xpgrd(npart),ygrd(npart),ypgrd(npart)
     &pgrd(npart),ejfvgrd(npart),sigmvgrd(npart),rvvgrd(npart),         &
     &dpsvgrd(npart),oidpsvgrd(npart),dpsv1grd(npart),   
xineff(npart),yineff(npart),              &
     &xpineff(npart),ypineff(npart)
```
In dbcommon:
```counteddpop(npart,numeffdpop)      , counted2d(npart,numeff,numeffdpop) part_hit(npart),part_abs(npart)
part_select(npart),nabs_type(npart)
part_impact(npart)
counted_r(npart,numeff),counted_x(npart,numeff),          &
     &counted_y(npart,numeff)
secondary(npart),tertiary(npart),other(npart),            &
     &part_hit_before(npart)
part_indiv(npart),part_linteract(npart)
                      ```

In +dbcollim:
```part_indiv(npart),part_linteract(npart)
x_in(npart),       &
     &xp_in(npart),y_in(npart),yp_in(npart),p_in(npart),s_in(npart),    &
     &indiv(npart),lint(npart),
impact(npart)
```

**Comment by kyrsjo** [comment]
Old standard-tracking lost particle code: If the aperture type is unset (abs(kp(ix)) not 2 or 3), then call `subroutine lostpar2`. Here it checks if x or y is greater than aper(1) or aper(2), or if x or y is NaN. In this case, set ilostch=1, and pstop(nlostp(j)) = true; nlostp(j) clearly acts as a pointer back to the original particle index.

For these particles, `aperv(nlostp(j),1) = apx(ix)` and `aperv(nlostp(j),2)` = apz(ix)`, `iv(nlostp(j)) = i`, `ixv(nlostp(j))=ix`, and 
```
          xvl(1,nlostp(j))=xv(1,j)
          xvl(2,nlostp(j))=xv(2,j)
          yvl(1,nlostp(j))=yv(1,j)
          yvl(2,nlostp(j))=yv(2,j)
          dpsvl(nlostp(j))=dpsv(j)
          ejvl(nlostp(j))=ejv(j)
          sigmvl(nlostp(j))=sigmv(j)
          numxv(nlostp(j))=numx
          nnumxv(nlostp(j))=numx
```

Then there is a writeout with the following format string:
```

10000 format(t10,'TRACKING ENDED ABNORMALLY'/t10, 'PARTICLE ',i7,       &
     &' RANDOM SEED ',i8, ' MOMENTUM DEVIATION ',g12.5/ t10,            &
     &' LOST IN REVOLUTION ',i8,' AT ELEMENT ',i4/ t10,                 &
     &'HORIZ:  AMPLITUDE = ',f15.3,'RE-APERTURE = ',f15.3/ t10,         &
     &'VERT:   AMPLITUDE = ',f15.3,'RE-APERTURE = ',f15.3/ t10,         &
     &'ELEMENT - LIST NUMBER ',i4,' TYP NUMBER ',i4,' NAME ',a16/)
      end
```

Finally the particle arrays are compressed:
```
     lnapx=napx
      do 30 j=napx,1,-1
        if(pstop(nlostp(j))) then
          if(j.ne.lnapx) then
            do 35 jj=j,lnapx-1
              jj1=jj+1
              nlostp(jj)=nlostp(jj1)
              xv(1,jj)=xv(1,jj1)
              xv(2,jj)=xv(2,jj1)
              yv(1,jj)=yv(1,jj1)
              yv(2,jj)=yv(2,jj1)
+if bnlelens
!GRDRHIC
!GRD-042008
              namepart(jj)=namepart(jj1)
!GRDRHIC
!GRD-042008
+ei
              dpsv(jj)=dpsv(jj1)
              sigmv(jj)=sigmv(jj1)
              ejfv(jj)=ejfv(jj1)
              ejv(jj)=ejv(jj1)
              rvv(jj)=rvv(jj1)
+if rvet
              rvet(jj)=rvet(jj1)
+ei
              oidpsv(jj)=oidpsv(jj1)
              dpsv1(jj)=dpsv1(jj1)
              clo6v(1,jj)=clo6v(1,jj1)
              clo6v(2,jj)=clo6v(2,jj1)
              clo6v(3,jj)=clo6v(3,jj1)
              clop6v(1,jj)=clop6v(1,jj1)
              clop6v(2,jj)=clop6v(2,jj1)
              clop6v(3,jj)=clop6v(3,jj1)
!--beam-beam element
              di0xs(jj)=di0xs(jj1)
              dip0xs(jj)=dip0xs(jj1)
              di0zs(jj)=di0zs(jj1)
              dip0zs(jj)=dip0zs(jj1)
              do 210 ib2=1,6
                do 210 ib3=1,6
                  tasau(jj,ib2,ib3)=tasau(jj1,ib2,ib3)
  210         continue
   35       continue
          endif
          lnapx=lnapx-1
        endif
   30 continue
 ```

It does *NOT* compress `al`, `as`, or `bl1v`, which are used in thick tracking. It also does *NOT* call `envarsv` to re-set them. This means that in the case of losses and momentum scan, the *thick* tracking will not be correct.

**Comment by rdemaria** [comment]
Some are fixed and they can be found in the bottom of the aperture module file. A check is needed to see if others have to be included.

**Comment by kyrsjo** [comment]
Tagging @rdemaria @vkbo @amereghe for final check  - I think the final stuff of this (like `ekv` being a copy of `ek`) is fixed in #870, but I think this one can be closed now.

**Comment by vkbo** [comment]
I think everything is covered now. Most of the thick arrays discussed here have been removed as they were only used as temporary arrays. The normalisation matrix is no longer per-particle, so the original issue is also solved.
